src/query/state.rs

```diff
@@ -0,0 +1,115 @@
+use crate::helpers::messaging::Gateway;
+
+use crate::protocol::QueryId;
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+/// The status of query processing
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[allow(dead_code)]
+pub enum QueryStatus {
+    /// Only query running on the coordinator helper can be in this state. Means that coordinator
+    /// sent out requests to other helpers and asked them to assume a given role for this query.
+    /// Coordinator is currently awaiting response from both peers.
+    Preparing,
+    /// Mesh network is established between helpers and they are ready to send and receive
+    /// messages
+    AwaitingInputs,
+    /// Helpers are negotiating the shared secrets to create PRSS and other things
+    Negotiating,
+    /// Query is being executed and can be interrupted by request.
+    Running,
+    /// Query processing has finished and the status of processing is available.
+    /// TODO: completion status and TTL
+    Completed,
+}
+
+impl<N> From<&QueryState<N>> for QueryStatus {
+    fn from(source: &QueryState<N>) -> Self {
+        match source {
+            QueryState::Preparing => QueryStatus::Preparing,
+            QueryState::AwaitingInputs(_, _) => QueryStatus::AwaitingInputs,
+        }
+    }
+}
+
+/// TODO: a macro would be very useful here to keep it in sync with `QueryStatus`
+pub enum QueryState<N> {
+    Preparing,
+    AwaitingInputs(N, Gateway),
+}
+
+impl<N> QueryState<N> {
+    pub fn transition(cur_state: Option<&Self>, new_state: Self) -> Result<Self, StateError> {
```

Comment:
It might be cleaner to have a default (or initial) value for the state than it would be to use an `Option<>`